File: 0090-subsets-ii/0090-subsets-ii.cpp
class Solution {
public:
set<vector<int>>s;
   void subset(vector<vector<int>>&ans,vector<int>&temp,vector<int>& nums,int i){
        
        if(i == nums.size()){
            sort(temp.begin(), temp.end());
            if(s.find(temp) == s.end()){
                 ans.push_back({temp});
                 s.insert(temp);
            }
           
            return;
        }
        
        temp.push_back(nums[i]);
        //Include
        subset(ans,temp,nums,i+1);

        //Backtracking part
        temp.pop_back();
         
        //Exclude
        subset(ans,temp,nums,i+1);
          int idx = i+1;
           while(idx<nums.size() && nums[idx] == nums[idx-1] ){
            idx++;
        }

     

    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
         vector<vector<int>>ans;vector<int>temp;int i = 0;
         sort(nums.begin(),nums.end());
         subset(ans,temp,nums,i);
         return ans;
    }
};

File: 0090-subsets-ii/0090-subsets-ii.java
class Solution {
    void help(int idx , int[] nums,  List<List<Integer>> ans,List<Integer>temp, Set<List<Integer>> st){
        if(idx == nums.length){
           
            st.add(new ArrayList<>(temp));
            return;
        }
        if(idx >= nums.length) return;

        temp.add(nums[idx]);
        help(idx+1, nums,ans ,temp,st);
        temp.remove(temp.size()-1);
        help(idx+1, nums,ans ,temp,st);
    }
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<Integer>temp = new ArrayList<>();
        List<List<Integer>>ans =  new ArrayList<>();
        Set<List<Integer>> st = new HashSet<>();
        help(0, nums, ans,temp,st);
        for(List<Integer> l: st){
         ans.add(l);
        }
        return ans;
    }
}